### DISCUSSION TITLE
Memory leaks in femtovg

### DISCUSSION THREAD

**Discussion by HitecExports** [slint-ui/slint#4079] [Q&A]
Hello!

Using Slint 1.3.2 / Windows 11 / FemtoVG

Substantial memory leaks occur when using repeated timer (triggers every 1ms for example) which renders a Text object.

Memory leakage appears with FemtoVG renderer (both debug and release builds).
Skia renderer seems to work ok.

```
slint::slint!(
    export component MainWindow inherits Window {

        width: 200px;
        height: 200px;

        callback tick();

        tick() => {
            root.elapsed-time += 1;
        }

        private property <int> elapsed-time;

        Text{
            text: elapsed-time;
        }
    }
);

pub fn main() {
    let main_window = MainWindow::new().unwrap();
    let timer = slint::Timer::default();
    {
        let main_window_weak = main_window.as_weak();
        timer.start(
            slint::TimerMode::Repeated,
            std::time::Duration::from_millis(1),
            move || {
                let main_window = main_window_weak.unwrap();
                main_window.invoke_tick();
            },
        );
    }
    main_window.run().unwrap();
}

```

In the example you can see:

Just after start the app consumes **22.4 Mb**

![image](https://github.com/slint-ui/slint/assets/19405195/80feff96-9e9d-4c97-a854-e6b272df433a)

After couple minutes of working the app consumes **41.7 Mb**

![image](https://github.com/slint-ui/slint/assets/19405195/5e4a89de-7cea-42a9-a2e7-377f2d1fadce)

I'm not creating any additional objects in the loop, so why is the application using more and more memory after each timer ticks?

**Comment by hunger** [comment]
It could be caching of data that we do internally, but I am not deep enough into that code to make any useful statement without investing serious effort.

I am afraid it will take a few days to get this answered: The people that wrote that code only return by the end of the week.

**Reply by HitecExports** [reply]
Thank you for the response!

For me it seems to be too much memory  for caching,  because this simple example could grow up to couple Gygabytes of memory usage after a while. But this memory leaks experienced with Femtovg. Skia works ok.

For now I will use Skia, because my application uses a timer and text redrawing quite a lot.

**Reply by hunger** [reply]
Great that it does at least not block you.

**Comment by tronical** [comment]
I'm been running your example on my Windows machine under Visual Studio's memory heap profiler. I observe two things:

1) The process memory usage grows initially and then stays flat at around 57MB. I don't observe a leak here. 

2) I've looked at various heap snapshots and the biggest memory allocations are:

  * From d3d11.dll when calling swap_buffers in opengl32.dll (which in turn calls into d3d11.dll)
  * From vm3dgl64.dll (Vmware GL driver?) via wglCreateContextAttribs.

I also observe that when I change your test case to make the window resizable, the size of the process memory usage depends on the size of the window. If I make the window bigger, the memory usage goes up, if I shrink it it goes down.

To me this looks similar to https://github.com/slint-ui/slint/discussions/3376 . I'm not entirely sure there's much that we can do on the Slint side about how the graphics drivers allocate memory. Unless there is a real leak, that is. Does the memory usage grow indefinitely for you, or does it flatten out?
